File: service/nodejs/receive.js
import amqp from 'amqplib/callback_api'
import { QUEUE_NAME, RABBITMQ_URL, PROJECT_ID, DEVICE_TOKEN } from './secret'
import key from './workshop-dekd.json' assert { type: 'json' }
import google from 'googleapis'
import https from 'https'

const HOST = 'fcm.googleapis.com'
const MESSAGING_SCOPE = 'https://www.googleapis.com/auth/firebase.messaging'
const SCOPES = [MESSAGING_SCOPE]
const PATH = '/v1/projects/' + PROJECT_ID + '/messages:send'

amqp.connect(RABBITMQ_URL, function (_, connection) {
    connection.createChannel(function (_, channel) {
        const queue = QUEUE_NAME
        channel.assertQueue(queue, { durable: false })

        console.log(
            '[*] Waiting for messages in %s. To exit press CTRL+C',
            queue
        )
        channel.consume(
            queue,
            function (msg) {
                console.log('[*] Received %s', msg.content.toString())
                const content = JSON.parse(msg.content.toString())
                sendFcmMessage({
                    message: {
                        token: DEVICE_TOKEN,
                        notification: {
                            title: content?.username,
                            body: content?.title,
                        },
                        data: content,
                    },
                })
            },
            {
                noAck: true,
            }
        )
    })
})

const getAccessToken = () => {
    return new Promise(function (resolve, reject) {
        const jwtClient = new google.Auth.JWT(
            key.client_email,
            null,
            key.private_key,
            SCOPES,
            null
        )
        jwtClient.authorize(function (err, tokens) {
            if (err) {
                reject(err)
                return
            }
            resolve(tokens.access_token)
        })
    })
}

const sendFcmMessage = (fcmMessage) => {
    getAccessToken().then(function (accessToken) {
        const options = {
            hostname: HOST,
            path: PATH,
            method: 'POST',
            headers: {
                Authorization: 'Bearer ' + accessToken,
            },
        }
        const request = https.request(options, function (resp) {
            resp.setEncoding('utf8')
            resp.on('data', function (data) {
                console.log(
                    '[*] Message sent to Firebase for delivery, response:'
                )
                console.log(data)
            })
        })
        request.on('error', function (err) {
            console.log('Unable to send message to Firebase')
            console.log(err)
        })
        request.write(JSON.stringify(fcmMessage))
        request.end()
    })
}

File: service/python/receive.py
import pika
import os
import firebase_admin
from firebase_admin import messaging
from firebase_admin import credentials
import secret
import json

cred = credentials.Certificate(os.getcwd() + '/'+ secret.credentials)
params = pika.URLParameters(secret.rabbitmqUrl)
connection = pika.BlockingConnection(params)
channel = connection.channel()
channel.queue_declare(queue=secret.queueName, passive=False,
                      durable=False, auto_delete=False)

firebase_admin.initialize_app(cred)

def callback(ch, method, properties, body):
    print('[*] Received: ' + str(body, 'UTF-8'))
    notifcation = json.loads(body.decode('utf-8'))
    sendNotification(notifcation['username'],
                     notifcation['title'],  notifcation['message'])
    if True:
        channel.basic_ack(delivery_tag=method.delivery_tag)


def sendNotification(token,username, title, messages):
    message = messaging.Message(
        notification=messaging.Notification(
            title=username,
            body=title
        ),
        data={
            'username': username,
            'title': title,
            'message': messages
        },
        token=secret.deviceToken,
    )
    response = messaging.send(message)
    print('[*] Successfully sent message: ', response)


channel.basic_consume(queue=secret.queueName,
                      on_message_callback=callback, auto_ack=False)

print('[*] Waiting for messages: ')
channel.start_consuming()
connection.close()
